tools/merge.py

```diff
@@ -0,0 +1,441 @@
+
+
+# weights of these layers will not be copied as they won't be valid after merging
+IGNORED_LAYERS = [
+    "optimizer",
+    "random_rng_state",
+    "np_rng_state",
+    "torch_rng_state",
+    "cuda_rng_state",
+    "rng_tracker_states",
+    "optimizer_state_dict",
+    "param_shapes",
+]
+
+
+def merge_checkpoints(
+    checkpoint_dir,
+    model_parallel_size,
+    pipe_parallel_size,
+    output_dir,
+    global_step=None,
+    parameter_mean=False,
+):
+    checkpoint_dir = Path(checkpoint_dir)
+    assert (
+        checkpoint_dir.is_dir()
+    ), f"checkpoint dir does not exist: {str(checkpoint_dir)}"
+    output_dir = Path(output_dir)
+
+    if global_step is None:
+        if os.path.isfile(checkpoint_dir / "latest"):
+            with open(checkpoint_dir / "latest") as f:
+                global_step = int(f.read().strip().replace("global_step", ""))
+        else:
+            raise ValueError("No global step provided")
+
+    weights_dir = checkpoint_dir / f"global_step{global_step}"
+
+    output_weights_dir = output_dir / f"global_step{global_step}"
+    output_configs_dir = output_dir / "configs"
+
+    print(f"* Merging from {weights_dir}", flush=True)
+    print(f"* Merging to {output_weights_dir}", flush=True)
+
+    # load modified configs and original mp size
+    config, orig_mp = get_output_config(
+        checkpoint_dir, output_dir, model_parallel_size, pipe_parallel_size
+    )
+
+    output_layer_parallelism = (
+        config.get("output_layer_parallelism")
+        or config.get("output-layer-parallelism")
+        or "row"
+    )
+
+    # this maps layer names to the dimension that should be merged
+    # the only layers that need to be merged are:
+    #   - vocab parallel embedding
+    #   - row parallel linear weights
+    #   - column parallel linear weights
+    #   - column parallel linear biases
+    # everything else is non parallel, and we just take the 0th mp index in this case
+    PARTITION_DIM_MAP = {
+        "word_embeddings.weight": 0,  # vocab
+        "attention.query_key_value.weight": 0,  # column
+        "attention.query_key_value.bias": 0,  # column
+        "mlp.dense_h_to_4h.bias": 0,
+        "mlp.dense_h_to_4h.weight": 0,
+        "attention.dense.weight": 1,  # row
+        "mlp.dense_4h_to_h.weight": 1,
+        # variable:
+        "final_linear.weight": 0 if output_layer_parallelism == "column" else 1,
+    }
+
+    # prepare output directories
+    if output_weights_dir.is_dir():
+        resp = input(
+            f"* Output weights dir ({output_weights_dir}) already exists. Do you want to overwrite it? (yes/no) "
+        )
+        if resp.lower() in ["yes", "y"]:
+            shutil.rmtree(output_weights_dir)
+        else:
+            exit()
+
+    for p in [output_weights_dir, output_configs_dir]:
+        p.mkdir(exist_ok=True, parents=True)
+
+    # save modified config
+    with open(output_configs_dir / "config.yml", "w") as f:
+        json.dump(config, f, indent=4)
```

Comment:
There may be a subtle bug here. In my original config I had a parameter like 
```
"eps": 1.0e-08
```
This script re-wrote that value to
```
"eps": 1e-08
```

Which led to the value being parsed as a string, not a float. Not sure why that's the case, but just FYI.

Reply:
Ugh. We ran into this issue before and I thought we had fixed it. It’s possible our “fix” was ultimately to write `1.0` everywhere though….

Reply:
Ah, i *think* if we save it out with yaml instead of json dump (it is a yaml after all) that should fix it. Bit of an oversight there, nice catch!